proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,280 @@
+---
+simd: '0023'
+title: Asynchronous Program Execution and Broadcast
+authors:
+  - Anatoly Yakovenko
+category: Standard/Meta
+type: Core
+status: Draft
+created: 2024-03-29
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This feature changes how the ledger is broadcast and executed. It
+separates proposing blocks full of user transactions from blocks
+with votes.  It allows for N concurrent builders of user transaction
+blocks, and it allows for asynchronous execution of user blocks.
+
+## Motivation
+
+1. Single leader for user transactions is a bottleneck.  Leaders
+are in a single spot in the world, while clients are all over the
+world. Latency is therefore based on how close a client is to a
```

Comment:
I think the idea is it reduces latency to nodes choosing which transactions make it into the block, not total confirmation time.

Let's re-use the locations in your example, and suppose I see a really great opportunity in some market.
Currently, you have to send to LA which will delay you 180+ms at least. At that point, it's more likely you missed out regardless of your priority fee because the leader will start packing their block as soon as possible so they can collect the most fees.

There's actually 2 separate aspects in this proposal that seem like they might help you get the opportunity:
1. Priority Re-ordering (could be done indepedent of 2)
2. Multiple Builders (seems dependent on 1)

Let's say we **just** have priority re-ordering without multiple leaders. This means if you pay a higher priority fee than anyone else going for the opportunity, you would get it...if limits were not already hit AND you made it in time. 180ms + delay is a significant fraction of the 400ms block-time, and it's possible (maybe likely) your tx doesn't make it into the leader's queue for consideration early enough to get included.

Having multiple builders increases your chances that one is closer to you make it to the closest one in time for them to include you in their user block and before limits are hit.

Not sure what @aeyakovenko intended here though - **because if this `UserBlock` isn't transmitted early enough then the builder misses the leader's slot and you still don't get included.** Maybe this only helps with the not hitting limits aspect, but you'd still have to be very early.

Reply:
What is the advantage of having multiple block builders separated from a single leader vs having multiple leaders who are also responsible for buiiding blocks?

Reply:
What do you mean by "multiple leaders"?
Is there a proposal describing this kind of setup?

Currently, a leader is tasked with producing the next block in the chain.
Ignoring forks, every block in the chain contains a single child block.
How would multiple parallel leaders produce a single subsequent block to be added to the chain?

Reply:
I think it's actually in the original whitepaper in section 4.4 https://github.com/solana-labs/whitepaper/blob/master/solana-whitepaper-en.pdf

Most of my understanding of how it could look like over the past year or so was informed from twitter discussions :) Basically, you'd need the bankless leader setup first, and from there you would go with an architecture similar to that described in the whitepaper.

Reply:
The multi leader approach will likely create more forks.  It's an attributable fault for a leader to produce two blocks for the same slot, but with multiple leaders if there is a turbine failure, the network has a similar partition but no one is at fault.

Reply:
Ahh I see. Makes sense.

I had assumed there was some benefit with having multiple leaders in the past.... I guess with single leaders what can we do to improve ddos resistance? I worry that there might be a high incentive in the future for competitors to try and "kill" solana. Maybe "sentry nodes" in front of leaders (?)